libraries/AP_Math/fourier.cpp

```diff
@@ -0,0 +1,220 @@
+		previous_index=_buffer_size-1;
+	
+	Vector3f signal_current=_signal[_buffer_index];//-_signal_mean/_buffer_size;
+	Vector2f fourier_coef;
+	
+	Vector2f new_result;
+	
+	fourier_coef.x=cosf(_phase[_buffer_index]);
+	fourier_coef.y=-sinf(_phase[_buffer_index]);
+	
+	signal_temp[0]=signal_current.x;
+	signal_temp[1]=signal_current.y;
+	
+	for(i=0;i<2;i++)
+	{
+		_fourier_transform[i].x -= (fourier_coef.x*signal_temp[i]);
+		_fourier_transform[i].y -= (fourier_coef.y*signal_temp[i]);
+	}
+	
+	actual_yaw=get_yaw_angle().x;
+	actual_heading=_phase[_buffer_index]-actual_yaw;
+	
+	_angle_tot.x -= _timing[_buffer_index].dt*(_timing[_buffer_index].omega.x*cosf(actual_heading) - _timing[_buffer_index].omega.y*sinf(actual_heading));
+	_angle_tot.y -= _timing[_buffer_index].dt*(_timing[_buffer_index].omega.y*cosf(actual_heading) + _timing[_buffer_index].omega.x*sinf(actual_heading));
+
+	
+	_signal_mean-=_signal[_buffer_index];
+	
+	_phase[_buffer_index]=_phase[previous_index]+new_timing->omega.z*new_timing->dt;
+	
+	if(_phase[_buffer_index]>M_PI)
+		_phase[_buffer_index]-=2*M_PI;
+	
+	new_timing->T=_timing[previous_index].T+new_timing->dt;
+
+	_signal_mean+=*new_sample;
+	
+	signal_current=*new_sample;//-_signal_mean/_buffer_size;
+	fourier_coef.x=cosf(_phase[_buffer_index]);
+	fourier_coef.y=-sinf(_phase[_buffer_index]);
+	
+	signal_temp[0]=signal_current.x;
+	signal_temp[1]=signal_current.y;
+	
+	for(i=0;i<2;i++)
+	{
+		_fourier_transform[i].x += (fourier_coef.x*signal_temp[i]);
+		_fourier_transform[i].y += (fourier_coef.y*signal_temp[i]);
+	}
+
+	actual_heading=_phase[_buffer_index]-actual_yaw;
+	
+	_angle_tot.x += _timing[_buffer_index].dt*(_timing[_buffer_index].omega.x*cosf(actual_heading) - _timing[_buffer_index].omega.y*sinf(actual_heading));
+	_angle_tot.y += _timing[_buffer_index].dt*(_timing[_buffer_index].omega.y*cosf(actual_heading) + _timing[_buffer_index].omega.x*sinf(actual_heading));
+
+	
+	_signal[_buffer_index]=*new_sample;
+	_timing[_buffer_index]=*new_timing;
+
+	_buffer_index++;
+
+	if(_buffer_index>=_buffer_size)
+		_buffer_index=0;
+	
+	new_result.x=get_pitch_angle();
+	new_result.y=get_yaw_angle().x;
+	
+	_last_result = _result_filter.apply(new_result,  new_timing->dt);
+	
+	if (_last_result.is_nan() || _last_result.is_inf()) {
+		_result_filter.reset();
+	}
+}
+
+void Fourier_Analysis::accumulate_discrete(Vector3f new_sample, float dt, Vector3f omega)
+{
+	Timing_Struct new_timing;
+
+	new_timing.dt=dt;
+	new_timing.omega=omega;
+	
+	this->accumulate(&new_sample, &new_timing);
+}
+
+float Fourier_Analysis::get_phase(void)
+{
+	int32_t last_index=_buffer_index-1;
+	
+	if(last_index<0)
+		last_index=_buffer_size-1;
+	
+	return _phase[last_index];
+}
+
+Vector2f Fourier_Analysis::get_result()
+{
+	return _last_result;
+}
+
+float Fourier_Analysis::get_pitch_angle(void){
+	int32_t next_index=_buffer_index+1;
+	
+	Vector2f result;
+	
+	if(next_index==_buffer_size)
+		next_index=0;
+
+	float scale=(float)_buffer_size/2.0;
```

Comment:
float divison is expensive. Instead of doing 5 of them here, can we turn them into multiplication?

Reply:
ok